Type checking support for signal forms

crisbeto opened the pull request:
Adds some logic to the template type checker to infer the expected type of nodes using the `Field` directive from the signal forms module, as well as some validations about the node that the `Field` directive is placed on.

crisbeto committed:
refactor(compiler-cli): allow the forms module to be loaded in tests

Updates the testing utilities to allow for `@angular/forms` to be loaded.

crisbeto committed:
fix(compiler-cli): infer types of signal forms set on native inputs

Sets up the logic for inferring the type of the signal form that is set on a native `input`, `textarea` or `select`.

crisbeto committed:
refactor(core): add utility type for extracting the value of a custom control

Adds the `ɵExtractFormControlValue` type that we can use during template type checking to extract the type of a custom control.

crisbeto committed:
fix(compiler-cli): infer type of custom field controls

Adds the logic to infer the type of a custom field control that's on the same element as the `Field` directive.

mmalerba commented on packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts:
```diff
@@ -436,6 +436,12 @@ export enum ErrorCode {
+  /** Raised when a form field directive is applied on an invalid node. */
+  FORM_FIELD_INVALID_NODE = 8022,
+
+  /** Raised when the user has an unsupported binding on a `Field` directive.a */
```
nit: directive.a

msmallest replied:
> directive.a

typo?

mmalerba commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -828,6 +842,44 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+        span,
+        ts.DiagnosticCategory.Error,
+        ngErrorCode(ErrorCode.FORM_FIELD_INVALID_NODE),
+        `This node is an invalid [field] directive host. The host must be a native form control ` +
```
There's also a third case I'd mention here, a component that provides an `NG_VALUE_ACCESSOR`

leonsenft replied:
So it's not possible to keep this diagnostic _and_ support `ControlValueAccessor` right? Because any valid use case of `ControlValueAccessor` would trigger this diagnostic?

Or is there any way to configure/support making this diagnostic conditional based on whether the project actually uses `ControlValueAccessor`?

mmalerba reviewed (commented):
I left a few comments about the documentation, for those I think we can document the correct behavior in the docstring, even if its not implemented yet in this PR, we can just leave a TODO next to it to finish implementing

crisbeto committed:
fix(compiler-cli): report invalid bindings on form controls

Adds validation that users don't bind to unsupported properties on nodes with the `Field` directive.

leonsenft commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -828,6 +842,44 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+        ngErrorCode(ErrorCode.FORM_FIELD_INVALID_NODE),
+        `This node is an invalid [field] directive host. The host must be a native form control ` +
+          `(such as <input>', '<select>', or '<textarea>') or a custom form control component with a ` +
+          `'value' or 'checked' model.`,
```
We did previously check that all inputs were signal-based, but I dropped that requirement. The instruction implementation doesn't care, since it uses `writeToDirectiveInput` which handles the disambiguation of the two.

It's still an open question as to whether we want to reintroduce the requirement simply to guide developers to use signal inputs?

leonsenft commented on packages/compiler-cli/test/ngtsc/signal_forms_spec.ts:
```diff
@@ -0,0 +1,306 @@
+      const diags = env.driveDiagnostics();
+      expect(diags.length).toBe(1);
+      expect(extractMessage(diags[0])).toBe(
+        `Type 'FieldTree<{ name: string; }, string | number>' is not assignable to type 'FieldTree<User, string | number>'.`,
```
Oh interesting, the type is `FieldTree` here (this is desired). Why is it different?

mmalerba replied:
I think because when the type is a primitive the child props stuff drops away and it simplifies down to just `() => FieldState<...>`, I think we can probably trick TS into not simplifying it by adding `& {}` or something to the `FieldTree` type in the primitive case.

leonsenft commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -741,6 +741,142 @@ class TcbGenericDirectiveTypeWithAnyParamsOp extends TcbDirectiveTypeOpBase {
+    const inputs = this.node instanceof TmplAstHostElement ? this.node.bindings : this.node.inputs;
+
+    for (const input of inputs) {
+      if (input.type === BindingType.Property && this.invalidBindingNames.has(input.name)) {
```
@alxhub and I discussed this and concluded that we probably want to prevent any flavor of setting these properties/attributes in the template.

* We absolutely do not want to support any dynamic binding, since that would "compete" with the field binding.
* At first it might seem reasonable to allow a static attribute as a "default" in case the bound field doesn't define a property, but this poses an issue in cases where the field binding itself is dynamic. If you bind one field that overrides the default, then swap to a field that doesn't define that property, we have to reset that property, and the original template attribute would be lost, so it doesn't really function as a default as you might expect.

So ultimately we probably want to raise this diagnostic for any variation of attribute/property of the invalid binding names.

crisbeto commented on packages/compiler-cli/test/ngtsc/signal_forms_spec.ts:
```diff
@@ -0,0 +1,306 @@
+      const diags = env.driveDiagnostics();
+      expect(diags.length).toBe(1);
+      expect(extractMessage(diags[0])).toBe(
+        `Type 'null' is not assignable to type '() => FieldState<string, string | number>'.`,
```
Yeah this is just what TS decided to set in the diagnostic.

crisbeto committed:
refactor(compiler-cli): split up form field ops

Splits up the custom form field and native form field into two separate `TcbOp`s with a shared base class.

crisbeto committed:
refactor(compiler-cli): flag unsupported static bindings on field directives

Expands the diagnostic for unuspported bindings on fields to also capture static attributes.

crisbeto committed:
refactor(compiler-cli): check that custom control conforms to control interfaces

Generates additional type checking code to ensure that custom control conform to either `FormValueControl` or `FormCheckboxControl`.

mmalerba commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -741,6 +741,263 @@ class TcbGenericDirectiveTypeWithAnyParamsOp extends TcbDirectiveTypeOpBase {
+/**
+ * Base class for `TcbOp`s that construct instances of the a signal form `Field`.
+ */
+abstract class TcbFieldDirectiveTypeBaseOp extends TcbOp {
```
Is there some way to see the output that this code generates?

crisbeto committed:
test(compiler-cli): set up TCB tests for signal forms

Adds TCB-specific tests for the generated code in signal forms since they tend to be a bit easier to read and follow.

mmalerba reviewed (approved):
reviewed-for: public-api

leonsenft commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -741,6 +741,257 @@ class TcbGenericDirectiveTypeWithAnyParamsOp extends TcbDirectiveTypeOpBase {
+
+    // `type` can't be bound, but is allowed as a static attribute.
+    this.unsupportedBindingFields = new Set([...commonUnsupportedNames, 'type']);
+    this.unsupportedAttributes = new Set([...commonUnsupportedNames, 'max-length', 'min-length']);
```
The `maxlength` and `minlength` HTML attributes aren't kebab case, so we shouldn't need these: https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Attributes/maxlength.

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -741,6 +741,257 @@ class TcbGenericDirectiveTypeWithAnyParamsOp extends TcbDirectiveTypeOpBase {
+
+    if (!(this.node instanceof TmplAstHostElement)) {
+      for (const attr of this.node.attributes) {
+        if (this.unsupportedAttributes.has(attr.name)) {
```
Attributes are case-sensitive in Angular's template AST.

leonsenft replied:
What does that mean? If a developer authors `mAxLEngth="5"` that will still render as `maxlength` in the DOM. Are you saying that Angular's AST/typechecker wouldn't/shouldn't catch this as if it were `maxlength`?

crisbeto replied:
At least as far as input matching goes, it's case insensitive, but I see your point about plain attributes. I've updated it with the `toLowerCase`.

mmalerba commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -741,6 +741,265 @@ class TcbGenericDirectiveTypeWithAnyParamsOp extends TcbDirectiveTypeOpBase {
+    super();
+
+    // Should be kept in sync with the `FormUiControl` bindings,
+    // defined in `packages/forms/signals/src/api/control.ts`.
```
can you use a similar trick as here https://github.com/angular/angular/pull/64590/files#diff-82d8054016d39f7efe27dd5dfd93dfb43287f6e17da2ba342e2aef54e8348596R687-R701 so that the type system can enforce keeping them in sync?
